java/ql/src/semmle/code/java/Member.qll

```diff
@@ -581,12 +581,13 @@ class Field extends Member, ExprParent, @field, Variable {
 
   /** Gets the initializer expression of this field, if any. */
   override Expr getInitializer() {
-    exists(AssignExpr e, InitializerMethod im |
+    exists(AssignExpr e, InitializerMethod im, ExprStmt exprStmt |
       e.getDest() = this.getAnAccess() and
       e.getSource() = result and
-      pragma[only_bind_out](result).getEnclosingCallable() = im and
-      // This rules out updates in explicit initializer blocks as they are nested inside the compiler generated initializer blocks.
-      pragma[only_bind_out](e.getEnclosingStmt().getParent()) = pragma[only_bind_out](im.getBody())
+      exprStmt.getExpr() = e and
+      // This check also rules out assignments in explicit initializer blocks
+      // (CodeQL models explicit initializer blocks as BlockStmt in initializer methods)
+      exprStmt.getParent() = im.getBody()
```

Comment:
Should I add `pragma[only_bind_out]` here again? If so to which of these predicate usages?

Reply:
I tried reproducing the bad join order observed in https://github.com/github/codeql/pull/6187, but the latest `codeql` doesn't produce that order any longer. We should just run a performance evaluation on this to see if it behaves badly.